src/main/clojure/clara/rules/engine.cljc

```diff
@@ -391,15 +391,74 @@
 
   (description [node] (str "QueryNode -- " query)))
 
+
+(defn node-rule-names
+  [children]
+  (loop [[next & rest] children flattened []]
+    (if next
+      (recur (into rest (:children next)) (conj flattened next))
+      (->> flattened
+           (keep :production)
+           (map :name)
+           (distinct)
+           (sort)))))
+
+(defn throw-constraint-exception
+  [error node fact env]
+  "Adds a useful error message when executing a constraint node raises an exception."
+  (let [children (:children node)
+        exp (:constraint-exp error)
+        rule-names (node-rule-names children)
+        condition (or (:condition node)
+                      (-> children first :condition))
+        type (:type condition)
+        constraints (or (:original-constraints condition)
+                        (:constraints condition))
+        condition-definition (into [type] constraints)
+        rule-header (if (> (count rule-names) 1)
+                      "of rules"
+                      "of rule")
+        unnamed-rule-msg (str "  An unnamed rule, provide names to your rules if"
+                              "you want them to be identified here.")
+        rule-list (->> rule-names
+                       (sort)
+                       (map #(if (nil? %) unnamed-rule-msg %))
+                       (map #(str "  " %))
+                       (clojure.string/join "\n"))
+        bindings (:bindings error)
+        bindings-description (if (empty? bindings)
+                               "with no bindings\n"
+                               (str "with bindings\n  " bindings))
+        message (clojure.string/join
+                 "\n"
+                 ["Exception raised during execution of constraint"
+                  (str "  " exp)
+                  "of condition"
+                  (str "  " condition-definition)
+                  rule-header
+                  rule-list
+                  "when processing fact"
+                  (str "  " (pr-str fact))
+                  bindings-description])]
+    (binding [*out* *err*] (println message))
+    (throw (Exception. message (:exception error)))))
```

Comment:
I don't think building the exception like this will work in ClojureScript; I think this relies on JVM interop.  If you look at the Travis build you'll see the message "WARNING: Use of undeclared Var clara.rules.engine/Exception at line 444 /home/travis/build/cerner/clara-rules/src/main/clojure/clara/rules/engine.cljc" and a similar one for \*err\*.

My suggestion is to use ex-info.  Basically, Clojure and ClojureScript define their own exception type that carries a payload of user-defined data.  This user-defined data can then be inspected in a REPL when the exception is thrown during development.   I created a quick gist demonstrating this functionality at https://gist.github.com/WilliamParker/d033cb89ea6ac996115e1712adf5a56a.

The docs on ex-info are at https://clojuredocs.org/clojure.core/ex-info